File: core/src/main/java/com/netflix/conductor/core/execution/tasks/Wait.java
package com.netflix.conductor.core.execution.tasks;

import java.time.Duration;
import java.util.Optional;

import org.springframework.stereotype.Component;

import com.netflix.conductor.core.execution.WorkflowExecutor;
import com.netflix.conductor.model.TaskModel;
import com.netflix.conductor.model.WorkflowModel;

import static com.netflix.conductor.common.metadata.tasks.TaskType.TASK_TYPE_WAIT;
import static com.netflix.conductor.model.TaskModel.Status.*;

@Component(TASK_TYPE_WAIT)
public class Wait extends WorkflowSystemTask {

    public static final String DURATION_INPUT = "duration";
    public static final String UNTIL_INPUT = "until";

    public Wait() {
        super(TASK_TYPE_WAIT);
    }

    @Override
    public void start(WorkflowModel workflow, TaskModel task, WorkflowExecutor executor) {
        task.setStatus(TaskModel.Status.IN_PROGRESS);
    }

    @Override
    public void cancel(WorkflowModel workflow, TaskModel task, WorkflowExecutor workflowExecutor) {
        task.setStatus(TaskModel.Status.CANCELED);
    }

    @Override
    public boolean execute(
            WorkflowModel workflow, TaskModel task, WorkflowExecutor workflowExecutor) {
        long timeOut = task.getWaitTimeout();
        if (timeOut == 0) {
            return false;
        }
        if (System.currentTimeMillis() > timeOut) {
            task.setStatus(COMPLETED);
            return true;
        }

        return false;
    }

    @Override
    public Optional<Long> getEvaluationOffset(TaskModel taskModel, long maxOffset) {
        if (taskModel.getWaitTimeout() > 0) {
            long seconds =
                    Duration.ofMillis(taskModel.getWaitTimeout() - System.currentTimeMillis())
                            .getSeconds();
            if (seconds == 0) {
                seconds = 1;
            }
            return Optional.of(seconds);
        }
        return Optional.empty();
    }

    public boolean isAsync() {
        return true;
    }
}

File: core/src/execution/tasks/Wait.ts
import { WorkflowSystemTask } from '../WorkflowSystemTask.js';
import type { WorkflowModel, TaskModel } from '../types.js';
import type { WorkflowExecutor } from '../WorkflowExecutor.js';
import { TaskType } from '@agentmesh/common';

export class Wait extends WorkflowSystemTask {
  constructor() {
    super(TaskType.WAIT);
  }

  override start(workflow: WorkflowModel, task: TaskModel, executor: WorkflowExecutor): void {
    task.status = 'IN_PROGRESS';
  }

  override cancel(
    workflow: WorkflowModel,
    task: TaskModel,
    workflowExecutor: WorkflowExecutor,
  ): void {
    task.status = 'CANCELED';
  }

  override execute(
    workflow: WorkflowModel,
    task: TaskModel,
    workflowExecutor: WorkflowExecutor,
  ): boolean {
    const timeOut = task.waitTimeout;
    if (timeOut === 0) {
      return false;
    }
    if (Date.now() > timeOut) {
      task.status = 'COMPLETED';
      return true;
    }
    return false;
  }

  override getEvaluationOffset(taskModel: TaskModel, maxOffset: number): number | undefined {
    if (taskModel.waitTimeout > 0) {
      const seconds = Math.max(1, Math.floor((taskModel.waitTimeout - Date.now()) / 1000));
      return seconds;
    }
    return undefined;
  }

  override isAsync(): boolean {
    return true;
  }
}
